### PR TITLE
AB-655 : mint with ERC20 / ERC20Permit

### PR DESCRIPTION
**Pull request by 0xPilou**
Added new NFT contract (ERC721ABLECoin) that features the following possibilities :
- Mint token with ETH
- Mint token with ERC20
- Mint token with ERC20Permit

### CODE REVIEW

File: src/token/ERC721/ERC721ABLECoin.sol [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    /**
+     * @notice
+     *  Withdraw ERC20 tokens from this contract to the caller
+     *  Only the contract owner can perform this operation
+     *
+     * @param _token token contract address to be withdrawn
+     * @param _amount amount to be withdrawn
+     */
+    function withdrawERC20(address _token, uint256 _amount) external virtual override onlyOwner {
+        if (_token == address(mintCurrency)) revert ABErrors.INVALID_PARAMETER();
```
Comment (jonathangus) [inline src/token/ERC721/ERC721ABLECoin.sol]: 🧠

File: src/token/ERC721/ERC721ABLECoin.sol [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+     * @notice
+     *  Mint `_quantity` tokens to `_to` address based on the current `_phaseId` if `_signature` is valid
+     *
+     * @param _to token recipient address (must be whitelisted)
+     * @param _phaseId current minting phase (must be started)
+     * @param _quantity quantity of tokens requested (must be less than max mint per phase)
+     * @param _signature signature to verify allowlist status
+     * @param _kycSignature signature to verify user's KYC status
+     */
+    function mintCoin(
```
Comment (jonathangus) [inline src/token/ERC721/ERC721ABLECoin.sol]: can we rename this to something else? could be mintWithToken/mintWithCoin/mintWithErc20 etc

**Review by jonathangus — COMMENTED** [review]
I think we should add this to ERC721AB or make it extenable 

```
class ERC721ABLE extends ERC721AB, ERC721ERC20Mint
```

also think we should extend phases to contain currency price

File: src/token/ERC721/ERC721ABLECoin.sol (hunk 2 above)
Comment (0xPilou) [reply src/token/ERC721/ERC721ABLECoin.sol]: renamed

**Comment by 0xPilou** [comment]
Update : 

- phase datatype now contains `priceETH` & `priceERC20` components
- ERC721AB now natively contains `acceptedCurrency` address state
- ERC721ABOE & ERC721ABLE now provide `mintWithERC20` & `mintWithERC20Permit` functions

Note : 
to block ERC20 payment, the `initDrop` function must be called with the param `_acceptedCurrency` equal to `0x0000000000000000000000000000000000000000` (the zero address)

**Review by asserules — APPROVED** [review]

**Pull request merged**